tools/extra/qpafilter/qpafilter.py

```diff
@@ -102,18 +106,75 @@ def write_sensor(self, sensor):
         upper_warning_id = self.threshold_map['Upper Warning']
         upper_fatal_id = self.threshold_map['Upper Fatal']
 
-        # TODO: endian-ness?
-        fmt = '<LLL'
-        self.outfile.write(struct.pack(fmt,
+        self.outfile.write(struct.pack(BLOB_STRUCT_FORMAT,
                                        sensor_id,
                                        upper_warning_id,
                                        warning))
-        self.outfile.write(struct.pack(fmt,
+        self.outfile.write(struct.pack(BLOB_STRUCT_FORMAT,
                                        sensor_id,
                                        upper_fatal_id,
                                        fatal))
 
 
+class blob_reader:
+    """Given an input file pointer, a sensor map, and a threshold map,
+       provide an iterator interface to reading the raw sensor data
+       items sequentially."""
+    def __init__(self, blobfp, sensor_map, threshold_map):
+        self.blob = blobfp.read()
+        self.start_marker_index = self.blob.find(BLOB_START_MARKER)
+        self.end_marker_index = self.blob.find(BLOB_END_MARKER)
+        self.sensor_map = sensor_map
+        self.threshold_map = threshold_map
+
+    class Iterator:
+        """Iterate over an in-memory blob. Calculate the number of
+           sensor entries by subtracting the offset of the first entry
+           from the offset of the end marker. Iterate and decode each
+           sensor entry, changing the raw data to human-readable output."""
+        def __init__(self,
+                     blob,
+                     start_marker_index,
+                     end_marker_index,
+                     sensor_map,
+                     threshold_map):
+            self.blob = blob
+            self.start_marker_index = start_marker_index
+            self.first_entry_index = self.start_marker_index + BLOB_STRUCT_SIZE
+            self.end_marker_index = end_marker_index
+            self.num_entries = ((self.end_marker_index -
+                                 self.first_entry_index) / BLOB_STRUCT_SIZE)
```

Comment:
do we also want to see if( end_marker_index-first_entry_index)%BLOB_STRUCT_SIZE is 0?

Reply:
I thought about that, but do you think it's necessary?  That would mean that some entity other than the tool created the blob, right?

Reply:
Probably not necessary. I was thinking of cases when the blob gets corrupted but we don't have to overcomplicate any more than this. I suppose checking for the start and end markers is sufficient.